duckduckgo/chrome-zeroclickinfo
# HTTPS Everywhere (Phase 2)

**Reviewer:**
@jdorweiler @russellholt 

**Depends on:** 
Internal PR #10337

## Description:
https://app.asana.com/0/263565713128358/394793249591636/f
This is Phase 2 of the HTTPS Everywhere implementation

## Steps to test this PR:
1. You'll need to change line 36 of /js/indexed-db.js to point at `lauren.duckduckgo.com` (vs prod) for server xhr call. _If you don't do this, server call will fail but extension should continue to function normally without HTTPS Everywhere upgrades._
2. Install extension via Firefox extension debugger.

To test simple http -> https upgrades:
1. navigate to `http://submit.pandora.com` -> *NOT `https://`
 see console output that verifies that extension upgraded the request to https. 
2. open popup, notice "forced secure connection" notice

To test whitelist:
1. Whitelist a site, confirm in console output that db rule check was skipped.

To test that we are dyamically whitelisting mixed content sites (that have both http and https content):
1. go to `/data/httpsWhitelist.json` and remove `adobe.com` entry
2. reload extension via ctrl+r in background debugger
3. navigate to http://adobe.com, watch address bar and console output. extension will attempt to load site as https, will get to end and reload tab automatically as http:// after whitelisting. from then on, https everywhere db check will be skipped for adobe.com

To test that we are handling redirects properly:
1. from terminal cli, navigate to `/test/mocks/redirect-server`. follow instructions in README for simulating an endless redirect loop locally. start redirect server.
2. with extension installed and running, navigate browser to `http://redirect.local:8070`. watch the console output and observe how redirects are only followed 7 times and then request is cancelled completely.



## Automated tests:

###### Reviewer Checklist:


###### PR Author Checklist:
